### PR TITLE
Node.js SDK : Error handling

### PR DESCRIPTION
**Pull request by benjaminstrasser**
- [x] SDK throws basic errors 
- [x] Create own centralized error handling with custom dagger errors 
  - [x] Replace basic errors with custom dagger SDK errors
  - [x] https://github.com/dagger/dagger/pull/3956#discussion_r1030628831 
  - [x] Others
- [x] Error documentation
  - [x] Documemntation via comments
  - [x] Add centralized error handling to docusaurus
- [x] ~~Optional: Amend documentation to adopt explicit error handling~~

### CODE REVIEW

**Comment by benjaminstrasser** [comment]
Thanks @slumbering  for the input and feedback!
> What about creating a new file `ErrorHandler.ts` or whatever you feel to centralize error handling ? A even cleaner way would be to create one file per error type. `GraphRequestError.ts` `EngineConnectError.ts` and so forth...
> 
Sound like a good plan. I will give it a try after work.

> Finally users can get a precise error:
> 
> ```ts
> connect(async (client: Client) => {
>     ....
> }).catch((error) => {
>   // We can then access to the error message or code, depending of what we want
>    console.error(error.message);
>    console.error(error.code);
> });
> ```
Here the question remains if we want to update the guides to reflect this possibility. I would personally add it but don't have particularly strong feelings about this.


About error references is see two options:
- use docusaurus and f.e. just add `errors.ts` https://github.com/dagger/dagger/blob/64a40a222640842b2f57c2829fd905c1db416aa3/website/docusaurus.config.js#L136-L154
- Write docs by hand, like f.e. [Angular Error List](https://angular.io/errors)

I would personally prefer option one and create a more detailed error reference once the SDK matures a bit more, but am open for input on this.

**Comment by slumbering** [comment]
> 



> Thanks @slumbering for the input and feedback!
> 
> > What about creating a new file `ErrorHandler.ts` or whatever you feel to centralize error handling ? A even cleaner way would be to create one file per error type. `GraphRequestError.ts` `EngineConnectError.ts` and so forth...
> 
> Sound like a good plan. I will give it a try after work.
> 
> > Finally users can get a precise error:
> > ```ts
> > connect(async (client: Client) => {
> >     ....
> > }).catch((error) => {
> >   // We can then access to the error message or code, depending of what we want
> >    console.error(error.message);
> >    console.error(error.code);
> > });
> > ```
> 
> Here the question remains if we want to update the guides to reflect this possibility. I would personally add it but don't have particularly strong feelings about this.
> 
> About error references is see two options:
> 
> * use docusaurus and f.e. just add `errors.ts` https://github.com/dagger/dagger/blob/64a40a222640842b2f57c2829fd905c1db416aa3/website/docusaurus.config.js#L136-L154
> * Write docs by hand, like f.e. [Angular Error List](https://angular.io/errors)
> 
> I would personally prefer option one and create a more detailed error reference once the SDK matures a bit more, but am open for input on this.

Agreed, option one is definitely what we want 👍, we do not want to maintain by hand ref docs.

**Comment by benjaminstrasser** [comment]
@slumbering 
Any feedback on the last commit? would love some feedback before starting the docs.
Was sticking with SDK errors only for now. How to interpret GraphQL errors is a little out of scope for this PR imo, but should probably be addressed.

**Comment by slumbering** [comment]
Awesome @benjaminstrasser thanks !
I'm going to have a look now 👀

File: sdk/nodejs/common/errors.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+  response: GraphQLResponse
+  request: GraphQLRequestContext
+}
+
+/*
+  TODO: ADD Description
+*/
+export class GraphQLRequestError extends DaggerSDKError {
+  name = "GraphQLRequestError"
+  code = "100"
```
Comment (slumbering) [inline sdk/nodejs/common/errors.ts:51]: Let's prefix the code with an `D` like `D100` (D stands for Dagger)

**Comment by slumbering** [comment]
I'm gonna push some changes in few minutes cc @benjaminstrasser

**Comment by slumbering** [comment]
@benjaminstrasser ! I've just push a commit. Basically i've created a more atomic structure with one error per file. I've also installed `node-color-log` to improve error readability in the terminal (no particular preference for this library feel free to change it if there is one that you prefer).

**Comment by benjaminstrasser** [comment]
@slumbering 
Hey. Thanks for the commit! I already started doing the basic typedoc documentation and will merge and push them shortly.
After that, imo we are ready for a final review.

**Comment by benjaminstrasser** [comment]
Hey @slumbering 
I manually merged most of your and my changes by hand as the automatic merge from intellij was a mess and the manual one was very easy. This gets squished anyway so I hope you forgive me for hacking on this one ;D

I added basic documentation to all errors and added them to docusaurus to autogenerate references. I will welcome any additions and feedback to the comments I wrote. Some of the people who wrote the code that errors could probably elaborate on them a lot better.

I also removed the automatic logging. Imo there is no need to log errors regardless of what happens to them. F.e in our case the error was logged during the test even though it was caught, which I find a little confusing.
I would much rather have the user choose to pretty print them if they desire to do so. This is why I created a `printStackTrace`method they can use and maybe we can add to the examples and such. Maybe I am wrong on this tho. Would love your feedback as well

I think this is it for this PR and I would love a review.

**Comment by slumbering** [comment]
Hey @benjaminstrasser ! I'm going to review this PR very soon :). We are going to ship some breaking changes this week and then the error handling can be next (next week or the week after). I let you know 👍

**Comment by slumbering** [comment]
Hey @benjaminstrasser ! I just rebased the main branch. Many changes has been done and I've all the context. That's why I did this rebase. We should be ready today for merging :).

File: sdk/nodejs/test/connect.spec.ts [hunk 2]
```diff
@@ -9,11 +10,28 @@ describe("NodeJS sdk", function () {
       const result = await client
         .container()
         .from("alpine")
-        .withExec(["apk", "add", "curl"])
-        .withExec(["curl", "https://dagger.io/"])
-        .stdout()
+        .exec(["apk", "add", "curl"])
```
Comment (slumbering) [inline sdk/nodejs/test/connect.spec.ts:13]:
```suggestion
        .withExec(["apk", "add", "curl"])
```

**Review by slumbering — CHANGES_REQUESTED** [review]

File: sdk/nodejs/common/errors/UnknownDaggerError.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { DaggerSDKError, DaggerSDKErrorOptions } from "./DaggerSDKError.js"
+
+/**
+ *  This error is thrown if the dagger SDK does not identify the error and just wraps the cause.
+ *  @see [UnknownDaggerError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.UnknownDaggerError)
```
Comment (slumbering) [inline sdk/nodejs/common/errors/UnknownDaggerError.ts:5]: The `@see` param links the page to itself.

File: sdk/nodejs/common/errors/DaggerSDKError.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
+import { log } from "../utils.js"
+
+export interface DaggerSDKErrorOptions {
+  cause?: Error
+}
+
+/**
+ * The base error. Every other error inherits this error.
+ * @see [DaggerSDKError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.DaggerSDKError)
```
Comment (slumbering) [inline sdk/nodejs/common/errors/DaggerSDKError.ts:9]: The `@see` param links the page to itself.
eg: https://deploy-preview-3947--devel-docs-dagger-io.netlify.app/current/sdk/nodejs/reference/classes/common_errors.DaggerSDKError

I guess we should remove it.

File: sdk/nodejs/common/errors/DockerImageRefValidationError.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { DaggerSDKError, DaggerSDKErrorOptions } from "./DaggerSDKError.js"
+
+interface DockerImageRefValidationErrorOptions extends DaggerSDKErrorOptions {
+  ref: string
+}
+
+/**
+ *  This error is thrown if the passed image reference does not pass validation and is not compliant with the
+ *  DockerImage constructor.
+ *  @see [DockerImageRefValidationError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.DockerImageRefValidationError)
```
Comment (slumbering) [inline sdk/nodejs/common/errors/DockerImageRefValidationError.ts:10]: The `@see` param links the page to itself.
I guess we should remove it.

File: sdk/nodejs/common/errors/EngineSessionPortParseError.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  parsedLine: string
+}
+
+/**
+ * This error is thrown if the EngineSession does not manage to parse the required port successfully.
+ * This can happen if
+ * - Reading the port times out after 30 seconds
+ * - The parsed port is not a number
+ * - the reader does not read a single line
+ * @see [EngineSessionPortParseError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.EngineSessionPortParseError)
```
Comment (slumbering) [inline sdk/nodejs/common/errors/EngineSessionPortParseError.ts:13]: The `@see` param links the page to itself.

File: sdk/nodejs/common/errors/GraphQLRequestError.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import { DaggerSDKError, DaggerSDKErrorOptions } from "./DaggerSDKError.js"
+
+interface GraphQLRequestErrorOptions extends DaggerSDKErrorOptions {
+  response: GraphQLResponse
+  request: GraphQLRequestContext
+}
+
+/**
+ *  This error originates from the dagger engine. It means that some error was thrown and sent back via GraphQL.
+ *  @see [GraphQLRequestError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.GraphQLRequestError)
```
Comment (slumbering) [inline sdk/nodejs/common/errors/GraphQLRequestError.ts:14]: The `@see` param links the page to itself.

File: sdk/nodejs/common/errors/InitEngineSessionBinaryError.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { DaggerSDKError, DaggerSDKErrorOptions } from "./DaggerSDKError.js"
+
+/**
+ *  This error is thrown if the dagger binary cannot be copied from the dagger docker image and copied to the local host.
+ *  @see [InitEngineSessionBinaryError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.InitEngineSessionBinaryError)
```
Comment (slumbering) [inline sdk/nodejs/common/errors/InitEngineSessionBinaryError.ts:5]: The `@see` param links the page to itself.

File: sdk/nodejs/common/errors/TooManyNestedObjectsError.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import { DaggerSDKError, DaggerSDKErrorOptions } from "./DaggerSDKError.js"
+
+interface TooManyNestedObjectsErrorOptions extends DaggerSDKErrorOptions {
+  response: unknown
+}
+
+/**
+ *  Dagger only expects one response value from the engine. If the engine returns more than one value this error is thrown.
+ *  @see [TooManyNestedObjectsError - Dagger.io](https://docs.dagger.io/current/sdk/nodejs/reference/classes/common_errors.TooManyNestedObjectsError)
```
Comment (slumbering) [inline sdk/nodejs/common/errors/TooManyNestedObjectsError.ts:9]: The `@see` param links the page to itself.

File: sdk/nodejs/common/errors/EngineSessionPortParseError.ts (hunk 6 above)
Comment (slumbering) [reply sdk/nodejs/common/errors/EngineSessionPortParseError.ts:13]: I guess we should remove it.

File: sdk/nodejs/common/errors/GraphQLRequestError.ts (hunk 7 above)
Comment (slumbering) [reply sdk/nodejs/common/errors/GraphQLRequestError.ts:14]: I guess we should remove it.

File: sdk/nodejs/common/errors/InitEngineSessionBinaryError.ts (hunk 8 above)
Comment (slumbering) [reply sdk/nodejs/common/errors/InitEngineSessionBinaryError.ts:5]: I guess we should remove it.

File: sdk/nodejs/common/errors/UnknownDaggerError.ts (hunk 3 above)
Comment (slumbering) [reply sdk/nodejs/common/errors/UnknownDaggerError.ts:5]: I guess we should remove it.

**Review by slumbering — COMMENTED** [review]
Tiny comments, nothing crazy. Once we're good, let's ship it 
🚢 🚀

**Comment by benjaminstrasser** [comment]
@slumbering 
Think I added all your feedback. Thanks for the review

**Comment by slumbering** [comment]
LGTM ! 🚀

**Pull request merged**